boot/bootutil/src/cbor.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (c) 2019 Linaro Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>  // TODO: Remove
+#include <stdlib.h>  // TODO: Remove
+
+#include "mcuboot_config/mcuboot_config.h"
+
+#ifdef MCUBOOT_SUIT
+
+#include <string.h>
+#include <inttypes.h>
+
+#include <bootutil/bootutil_log.h>
+
+MCUBOOT_LOG_MODULE_DECLARE(mcuboot);
+
+#include "cbor.h"
+
+/* Decode the initial byte, and the associated value.  Returns a
+ * non-zero error upon error.  On 0 for success, 'major' will be
+ * written with the major value (0-7), and minor will be written with
+ * the minor value.  This will consider indefinite values, and 64-bit
+ * values to be invalid.  On success the slice will then point to the
+ * following item after the tag.  The slice is not modified on error.
+ */
+static int
+get_cbor_tag(struct slice *data,
+             uint8_t *major,
+             uint32_t *minor)
+{
+    if (data->len < 1) {
+        return -1;
+    }
+
+    uint8_t first = *data->base;
+
+    uint8_t maj = first >> 5;
+    uint8_t ext = first & 0x1f;
+
+    unsigned want;
+    if (ext < 24) {
+        *major = maj;
+        *minor = ext;
+        data->base++;
+        data->len--;
+        return 0;
+    } else if (ext == 24) {
+        want = 1;
+    } else if (ext == 25) {
+        want = 2;
+    } else if (ext == 26) {
+        want = 4;
+    } else {
+        /* If we want uint64 support, add it above here. */
+        return -1;
+    }
```

Comment:
*maybe* would gain a few bytes by doing:
```
    ...
    return 0;
} else if (ext > 26) {
    return -1;
}

want = 1 << (ext - 24);
```

Reply:
10-12 bytes. 10 bytes of code difference, possibly causing 12 bytes of padding.